pkg/tls/ca.go

```diff
@@ -49,97 +45,167 @@ type CA struct {
+	t.Subject = pkix.Name{CommonName: name}
+	t.IsCA = true
+	t.MaxPathLen = -1
+	t.BasicConstraintsValid = true
+	t.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageCRLSign
+	der, err := x509.CreateCertificate(rand.Reader, &t, &t, ca.PrivateKey.Public(), ca.PrivateKey)
+	if err != nil {
+		return
+	}
+	crt, err := x509.ParseCertificate(der)
+	if err != nil {
+		return
+	}
 
-// TrustAnchorPEM returns the PEM-encoded X.509 certificate of the trust anchor
-// (root CA).
-func (ca *CA) TrustAnchorPEM() string {
-	return ca.rootPEM
+	ca.Crt.Certificate = crt
+	return
 }
 
-// IssueEndEntityCertificate creates a new certificate that is valid for the
+// GenerateIntermediary creates a new certificate that is valid for the
 // given DNS name, generating a new keypair for it.
-func (ca *CA) IssueEndEntityCertificate(dnsName string) (*CertificateAndPrivateKey, error) {
+func (ca *CA) GenerateIntermediary(name string, maxPathLen int) (*CA, error) {
 	privateKey, err := generateKeyPair()
 	if err != nil {
 		return nil, err
 	}
-	p8, err := x509.MarshalPKCS8PrivateKey(privateKey)
+
+	t := ca.createTemplate(&privateKey.PublicKey)
+	t.Subject = pkix.Name{CommonName: name}
+	t.IsCA = true
+	t.MaxPathLen = maxPathLen
+	t.BasicConstraintsValid = true
+	t.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageCRLSign
+	crtb, err := x509.CreateCertificate(rand.Reader, &t, ca.Crt.Certificate, privateKey.Public(), ca.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+	crt, err := x509.ParseCertificate(crtb)
 	if err != nil {
 		return nil, err
 	}
 
-	template := ca.createTemplate(&privateKey.PublicKey)
-	template.DNSNames = []string{dnsName}
-	crt, err := x509.CreateCertificate(rand.Reader, &template, ca.root, &privateKey.PublicKey, ca.privateKey)
+	ica := newCA()
+	ica.validity = ca.validity
+	ica.clockSkewAllocance = ca.clockSkewAllocance
+	ica.PrivateKey = privateKey
+	ica.Crt.Certificate = crt
+	ica.Crt.TrustChain = append(ca.Crt.TrustChain, ca.Crt.Certificate)
+	return ica, nil
+}
+
+// GenerateEndEntity creates a new certificate that is valid for the
+// given DNS name, generating a new keypair for it.
+func (ca *CA) GenerateEndEntity(dnsName string) (*EndEntity, error) {
+	pk, err := generateKeyPair()
 	if err != nil {
 		return nil, err
 	}
-	return &CertificateAndPrivateKey{
-		Certificate: crt,
-		PrivateKey:  p8,
-	}, nil
+	crt, err := ca.SignEndEntity(&x509.CertificateRequest{
+		Subject:   pkix.Name{CommonName: dnsName},
+		DNSNames:  []string{dnsName},
+		PublicKey: &pk.PublicKey,
+	})
+	return &EndEntity{PrivateKey: pk, Crt: crt}, err
 }
 
-// createTemplate returns a certificate template for a non-CA certificate with
-// no subject name, no subjectAltNames. The template can then be modified into
-// a (root) CA template or an end-entity template by the caller.
+// SignEndEntity creates a new certificate that is valid for the
+// given DNS name, generating a new keypair for it.
+func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (Crt, error) {
```

Comment:
And finally I'd also have this func return `*Crt`, and then I'd embed `*Crt` in the `EndEntity` struct.

```diff
diff --git a/pkg/tls/ca.go b/pkg/tls/ca.go
index f157558f..c9f3cc5b 100644
--- a/pkg/tls/ca.go
+++ b/pkg/tls/ca.go
@@ -53,7 +53,7 @@ type Crt struct {
 
 type EndEntity struct {
 	PrivateKey *ecdsa.PrivateKey
-	Crt
+	*Crt
 }
 
 func newCA() *CA {
@@ -175,10 +175,10 @@ func (ca *CA) GenerateEndEntity(dnsName string) (*EndEntity, error) {
 
 // SignEndEntity creates a new certificate that is valid for the
 // given DNS name, generating a new keypair for it.
-func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (Crt, error) {
+func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (*Crt, error) {
 	pubkey, ok := csr.PublicKey.(*ecdsa.PublicKey)
 	if !ok {
-		return Crt{}, fmt.Errorf("CSR must contain an ECDSA public key: %+v", csr.PublicKey)
+		return nil, fmt.Errorf("CSR must contain an ECDSA public key: %+v", csr.PublicKey)
 	}
 
 	t := ca.createTemplate(pubkey)
@@ -193,16 +193,16 @@ func (ca *CA) SignEndEntity(csr *x509.CertificateRequest) (Crt, error) {
 
 	crtb, err := x509.CreateCertificate(rand.Reader, &t, ca.Crt.Certificate, pubkey, ca.PrivateKey)
 	if err != nil {
-		return Crt{}, fmt.Errorf("Failed to create certificate: %s", err)
+		return nil, fmt.Errorf("Failed to create certificate: %s", err)
 	}
 
 	crt, err := x509.ParseCertificate(crtb)
 	if err != nil {
-		return Crt{}, fmt.Errorf("Failed to parse certificate: %s", err)
+		return nil, fmt.Errorf("Failed to parse certificate: %s", err)
 	}
 
 	chain := append(ca.Crt.TrustChain, ca.Crt.Certificate)
-	return Crt{Certificate: crt, TrustChain: chain}, nil
+	return &Crt{Certificate: crt, TrustChain: chain}, nil
 }
 
 // createTemplate returns a certificate t for a non-CA certificate with
```